src/lem/constrainer.rs

```diff
@@ -44,14 +44,50 @@ impl<F: LurkField> AllocationManager<F> {
             }
         }
     }
+
+    /// Calls `get_or_alloc_num` to allocate tag and hash for a pointer.
+    pub(crate) fn get_or_alloc_ptr<CS: ConstraintSystem<F>>(
+        &mut self,
+        cs: &mut CS,
+        z_ptr: &ZPtr<F>,
+    ) -> Result<AllocatedPtr<F>> {
+        Ok(AllocatedPtr::from_parts(
+            &self.get_or_alloc_num(cs, z_ptr.tag.to_field())?,
+            &self.get_or_alloc_num(cs, z_ptr.hash)?,
+        ))
+    }
 }
 
+enum HashArity {
+    A2,
+    A3,
+    A4,
+}
+
+/// Information needed to constrain each hash slot
+type SlotsData<F> = Vec<(
+    // Arity of the hash
+    HashArity,
+    // Variable that indicates if we are in a concrete path or not
+    Boolean,
+    // Hash preimage
+    Vec<AllocatedNum<F>>,
+    // Hash value
+    AllocatedNum<F>,
+    // Allocated pointer name containing the result of the hash
+    String,
+)>;
```

Comment:
Why not use a struct with meaningful names? These large tuples or arbitrary data that depend on code comments for any hope of interpretation just make using and reading the code harder, in my opinion. If you use sensible field names, you can still keep building and destructuring compact in the using code.